### PR TITLE
New packages for e2e tests with sample test.

### PR DESCRIPTION
**Pull request by alanconway**
TLDNR; Look at [example of new forwarding test ](https://github.com/alanconway/cluster-logging-operator/blob/new-e2e-tests/test/e2e/logforwarding/fluent/fluent_test.go#L1)

Goals:
- Faster tests: run cluster actions concurrently, report slow actions.
  - New fluentd forwarding tests < 2 min, old tests ~ 5 min.
- Only pre-test requirement is `make deploy`
- Don't fail if resources are left behind by other tests, delete and recreate.
- Passing tests delete their resources, failing tests leave them for debugging.
- Labels for easy cleanup: `oc delete all -A -l test-client`

Package summary:

test
- UniqueName: strong crypto random IDs to avoid collisions.
- CmdReader: tail output of `oc exec` commands - read instead of polling.
- FailGroup: paralellize Ginkgo assertions safely.
- UnlessFailed: clean-up only on pass, leave evidence for debugging on fail.

test/runtime
- generic manipulation of objects via the runtime.Object interface.
- easy constructors for corev1 and loggingv1 objects.
- Decode objects from YAML (also test.MustUnmarshal to fill in parts of objects.)

test/client:
- Generic client for any object struct type using runtime.Object interface.
- Client.Watch: watch for changes instead of polling.
- Recreate: Delete+create - no more "already exists" failures.
- Process-wide singleton client (multiple client creation is slow)
- Auto-label resources for easy clean-up: `oc delete all -A -l test-client`
- CreateUniqueNS: unique namespace per test

test/client/fluentd:
- Flexible fluentd receiver, any number of listening ports.
- Tail output files instead of polling
- TLS secured sources.

test/certificate:
- Simplified certificate management for tests.

test/e2e/logforwarding/fluentnew:
- Sample test using the above (will eventually replace existing fluent test)
- Validates log data, doesn't just check for existence.
- Paralellizes operations where possible using FailGroup.

### CODE REVIEW

File: test/cmdreader_test.go
Comment (vimalk78) [inline test/cmdreader_test.go:14]: focussed test

**Comment by alanconway** [comment]
/hold
for comments

File: test/client/client_suite_test.go
Comment (periklis) [inline test/client/client_suite_test.go:12]: IIRC correct we configure the junit reporter for E2E tests. I suggest to consider the similar ginkgo register routines in our E2E test suite.

File: test/client/client_test.go
Comment (periklis) [inline test/client/client_test.go:29]: Does this imply that we keep the namespace if a failure occurs? Is this required for a test must-gather later?

File: test/client/client_test.go
Comment (periklis) [inline test/client/client_test.go:41]: Typo? Shouldn't this be `recreates`?

File: test/client/client_test.go
Comment (periklis) [inline test/client/client_test.go:14]: Isn't `defer ginkgo.GinkgoRecover()` required for a testcase that makes use of go routines?

File: test/client/fluentd/fluentd_suite_test.go
Comment (periklis) [inline test/client/fluentd/fluentd_suite_test.go:12]: IIRC correct we configure the junit reporter for E2E tests. I suggest to consider the similar ginkgo register routines in our E2E test suite.

File: test/client/fluentd/receiver_test.go
Comment (periklis) [inline test/client/fluentd/receiver_test.go:16]: Isn't defer ginkgo.GinkgoRecover() required for a testcase that makes use of go routines?

File: test/client/fluentd/receiver_test.go
Comment (periklis) [inline test/client/fluentd/receiver_test.go:17]: The whole suite is currently focus on this single test case!!!

File: test/cmdreader.go
Comment (periklis) [inline test/cmdreader.go:15]: I suggest to rename `DefaultTimeout` to `DefaultSuccessTimeout` and `ShortTimeout` to `DefaultFailureTimeout`

File: test/cmdreader_test.go
Comment (periklis) [inline test/cmdreader_test.go:13]: Same here missing `ginkgo.GinkgoRecover()`?

File: test/e2e/logforwarding/fluentnew/fluentnew_suite_test.go
Comment (periklis) [inline test/e2e/logforwarding/fluentnew/fluentnew_suite_test.go:12]: Same here. Maybe missing the junit reporter.

File: test/e2e/logforwarding/fluentnew/fluentnew_test.go
Comment (periklis) [inline test/e2e/logforwarding/fluentnew/fluentnew_test.go:15]: `object` is quite a generic pkg name. Why not calling it `crds`?

File: test/e2e/logforwarding/fluentnew/fluentnew_test.go
Comment (periklis) [inline test/e2e/logforwarding/fluentnew/fluentnew_test.go:26]: Same here missing ginkgo.GinkgoRecover() because we use go routines?

File: test/client/client_suite_test.go
Comment (alanconway) [reply test/client/client_suite_test.go:12]: Yep. The J2EE reporter and the collection of debugging artifacts should also move to Go in a later PR.

File: test/client/client_test.go
Comment (alanconway) [reply test/client/client_test.go:29]: It does, that's something to be considered. When I'm running tests one-by-one, it's much easier if the actual namespace is still around. When we're running on CI I'm not sure if that's helpful or not. We might want to configure this based on whether we're in CI or not, or maybe it's just OK to leak these resources since they don't interfere with other tests and can be cleaned up with an `oc delete -l test-client` We'll see how it works in practice.

File: test/client/client_test.go
Comment (alanconway) [reply test/client/client_test.go:41]: yep

File: test/client/fluentd/fluentd_suite_test.go
Comment (alanconway) [reply test/client/fluentd/fluentd_suite_test.go:12]: +1

File: test/cmdreader.go
Comment (alanconway) [reply test/cmdreader.go:15]: I like it.

Comment (alanconway) [reply test/cmdreader.go:15]: I despise the very notion of a "DefaultFailureTimeout" but I haven't figured out how to eliminate it yet. It's "wishful thinking" testing - "well I didn't look very hard and I didn't find anything, so it must not be there."

File: test/client/client_test.go
Comment (alanconway) [reply test/client/client_test.go:14]: It's required in the goroutine. The FailGroup type takes care of this for you so you can safely say:
```
g := test.FailGroup
g.Go(func() { Expect(blah).To(BeBoring()); }
g.Go(func() { Expect(geewhiz).To(BeGreatlyExciting()); }
g.Wait()
```
If you start goroutines by some other method then you would need the defer ginkgo stuff.

File: test/e2e/logforwarding/fluentnew/fluentnew_test.go
Comment (alanconway) [reply test/e2e/logforwarding/fluentnew/fluentnew_test.go:15]: Because corev1.Pod etc. are not CRDs, and because the central type in the API is called runtime.Object. I'm open to suggestions though, I had trouble deciding on this this. It's inconsistent as it stands because I gave clusterlogging a package of its own but no the other types.

My first take was:

```
package test
type Object struct{ o runtime.Object }
func (o Object) NamespacedName() ....
```

So you can say:

```
var o runtime.Object
obj := test.Object(o)
m := obj.Meta()
nn := obj.NamespacedName()
```

I was pretty pleased with myself, but in practice I almost always ended up only needing to call one method so it was usually:

```
nn := test.Object(o).NamespacedName()
```

in which case a simple package function is, well, simpler. I thought I might be getting fancy for no real gain (I do that sometimes) What do you think? I could be persuaded to go back to that, or if you have something simpler in mind...

Comment (periklis) [reply test/e2e/logforwarding/fluentnew/fluentnew_test.go:15]: I suggest to improve identifying speed and lessen cognitive bias between apimachinery are code, why not replicate the package name as `runtime`. That buys us:
```
github.com/openshift/cluster-logging-operator/test/runtime/clusterlogging
```
which is easier to relate to
```
"k8s.io/apimachinery/pkg/runtime"
```
WDYT?

Comment (alanconway) [reply test/e2e/logforwarding/fluentnew/fluentnew_test.go:15]: +1 I like it

File: test/cmdreader_test.go
Comment (alanconway) [reply test/cmdreader_test.go:14]: Fixed

File: test/e2e/logforwarding/fluentnew/fluentnew_test.go
Comment (alanconway) [reply test/e2e/logforwarding/fluentnew/fluentnew_test.go:15]: Fixed

**Review by periklis — APPROVED** [review]
Looks good to me so far. I will leave it to others to review and apply the labels.

File: Dockerfile
Comment (jcantrill) [inline Dockerfile:1]: Should probably do this rebased from https://github.com/openshift/cluster-logging-operator/pull/736

File: README.md
Comment (jcantrill) [inline README.md:25]: Remove since this no longer exists and was devpreview api

File: README.md
Comment (jcantrill) [inline README.md:27]: @mburke5678 is there a live like to 4.6?

File: bundle/manifests/logging.openshift.io_clusterlogforwarders_crd.yaml
Comment (jcantrill) [inline bundle/manifests/logging.openshift.io_clusterlogforwarders_crd.yaml:16]: Isn't this part of a PR already merged?

File: cmd/manager/main.go
Comment (jcantrill) [inline cmd/manager/main.go:48]: Already part of the log PR refactoring which is merged?

File: docs/HACKING.md
Comment (jcantrill) [inline docs/HACKING.md:50]: its not 'requires'?

File: hack/get-openshift-client.sh
Comment (jcantrill) [inline hack/get-openshift-client.sh:11]: merged PR?

**Review by jcantrill — CHANGES_REQUESTED** [review]

**Comment by alanconway** [comment]
@jcantrill are you ready to LGTM this?

File: Dockerfile
Comment (alanconway) [reply Dockerfile:1]: Done

Comment (alanconway) [reply Dockerfile:1]: Done.

File: README.md
Comment (alanconway) [reply README.md:25]: This file wasn't part of the PR, but fixed anyway/

File: README.md
Comment (alanconway) [reply README.md:27]: Fixed - removed the version specific reference, people can find the latest from the home page.

File: cmd/manager/main.go
Comment (alanconway) [reply cmd/manager/main.go:48]: Didn't touch this file. After rebase it matches latest master.

File: bundle/manifests/logging.openshift.io_clusterlogforwarders_crd.yaml
Comment (alanconway) [reply bundle/manifests/logging.openshift.io_clusterlogforwarders_crd.yaml:16]: Not sure what changes got picked up here. I didn't touch that file in this PR. Have a look at the rebase.

File: hack/get-openshift-client.sh
Comment (alanconway) [reply hack/get-openshift-client.sh:11]: Again, didn't touch - check rebased PR

**Comment by alanconway** [comment]
@jcantrill figured out the diff confusion, this was still targetted at master-post-release. Switched to master should make more sense now.

File: docs/HACKING.md
Comment (alanconway) [reply docs/HACKING.md]: Merge confusion. Check again.

File: test/helpers/fluentd/secret.tmp
Comment (jcantrill) [inline test/helpers/fluentd/secret.tmp:1]: should this be deleted?

File: test/runtime/logger.go
Comment (jcantrill) [inline test/runtime/logger.go:14]: Can we call this LogGenerator or the like and rename the file.  I think "Logger" i think of logging my error messages or the link

**Review by jcantrill — CHANGES_REQUESTED** [review]

File: test/client/client.go
Comment (alanconway) [reply test/client/client.go:1]: It's a self-contained unit of functionality that can be re-used in other circumstances so I think it makes sense for it to be a package.

File: test/client/client.go
Comment (alanconway) [reply test/client/client.go:47]: Nope, go map is pointer-like (as are slice, array and chan) so a copy of a map points to the same set of key/value pairs as the original.

File: test/client/client.go
Comment (alanconway) [reply test/client/client.go:36]: Nope, return values must all be named or all be nameless.

File: test/client/client.go
Comment (alanconway) [reply test/client/client.go:120]: I came across this https://github.com/kubernetes/kubernetes/issues/42594#issuecomment-285801993 which says "The current state is that kubectl delete always waits (for ex: it waits for pods to be deleted when running kubectl delete rs or kubectl delete deployment). Namespace deletion and pod graceful deletion are exceptions." However that's not a very definitive reference. I'll beef up Recreate so it does a Watch() for the delete event.

File: test/helpers/fluentd/secret.tmp
Comment (alanconway) [reply test/helpers/fluentd/secret.tmp:1]: Yes, done.

File: test/runtime/logger.go
Comment (alanconway) [reply test/runtime/logger.go:14]: Done.

File: test/e2e/logforwarding/fluent/fluent_secure_test.go
Comment (alanconway) [reply test/e2e/logforwarding/fluent/fluent_secure_test.go:56]: Fair point, I'll make it a method on Source.

File: test/const.go
Comment (alanconway) [reply test/const.go:25]: :) circular dependency is exactly what happend. Since package test doesn't have any dependencies, I can re-define helpers/constants.go in terms of the test/const.go so there's only one definition.

File: test/cmdreader.go
Comment (alanconway) [reply test/cmdreader.go:1]: Done.
Can you summarize what you think belongs in package test vs. in a package under helpers? I can drop in a README for future devs adding new stuff.

File: test/certificate/certificate.go
Comment (alanconway) [reply test/certificate/certificate.go:1]: helpers it is.

**Comment by alanconway** [comment]
@jcantrill I think I've addressed all your comments. LGTM?

File: test/client/client.go
Comment (jcantrill) [reply test/client/client.go:1]: I wasn't suggesting it should not be a package but that we may have other clients that could logically be in their own packages:
```
test/client/openshift/client.go
test/client/podman/client.go
```

File: test/client/test.go
Comment (jcantrill) [reply test/client/test.go:1]: Still open question

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (jcantrill) [inline test/e2e/logforwarding/fluent/fluent_test.go:65]: As a team, we had a conversation of terse vs explicit and frankly I find this syntax non-intuitive. Is there something we can do here by name that can make it more obvious what it does?  I don't understand why its structured this way at all

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (jcantrill) [inline test/e2e/logforwarding/fluent/fluent_test.go:75]: Is this something that can be automatically done for us via a helper method?

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (jcantrill) [inline test/e2e/logforwarding/fluent/fluent_test.go:76]: Maybe more verbose of what I'm getting?   "ReaderForSourceType(" ??

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (alanconway) [reply test/e2e/logforwarding/fluent/fluent_test.go:65]: How would you like it structured? Here are some other options
```
// Use struct field names to be more explicit
receiver.AddSource("name", Source{Port: 24224, Type: "forward"})
// Provide all sources as map on creation, no Add function
receiver := NewReceiver(map[string]*Source{ "name": Source{Port: 24224, Type: "forward"}, .... }
```
I used "AddSource/Source" because that's what fluentd calls it, personally I would tend to use "Listen/Listener" to follow the go net library. I guess it depends if we want to follow fluent or Go conventions here.

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (alanconway) [reply test/e2e/logforwarding/fluent/fluent_test.go:75]: We could mke the WithCancel line shorter with a terse function name, but I think that's a bad case of choosing terse over explicit. 

Possibly you could persuade me to take a different approach entirely - the idea here is that we are expecting logs from some sources, and not from others. Therefore we wait (forever) on the sources where we expect nothing, and we cancel those waits when we've got everything we expected from the other sources.

A simpler alternative would be to wait with a fixed timeout where we expect nothing, then no need for cancellation. I don't like fixed timeouts because they are always too long or too short, and that's one of the reasons our tests currently spend a lot of time sleeping for nothing. Using the wait time for expected results seems like a nice natural timeout based real conditions when the test runs.

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (alanconway) [reply test/e2e/logforwarding/fluent/fluent_test.go:76]: Done.

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (alanconway) [reply test/e2e/logforwarding/fluent/fluent_test.go:65]: Updated using explicit struct fields, I like this better than the original. Whaddya think @jcantrill ?

Comment (jcantrill) [reply test/e2e/logforwarding/fluent/fluent_test.go:65]: The source was less of a concern to me then `t.group.Go(func() { ExpectOK(t.receiver.Create(t.Client)) })` which doesn't read well to be or make it obvious whats going on.   I'm coming from the perspective of what is it I need to do when I add a test and why am I doing it

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (jcantrill) [reply test/e2e/logforwarding/fluent/fluent_test.go:75]: I'm not saying we should not do this or this is not correct.  I'm trying to emphasize the "why"?  Why am I doing this and if I'm writing a test is it going to be obvious that I forgot to add it.  Can the framework/test design make it more obvious what to  do and why by either:
* making a name that makes it explicit
* Adding a function that "does it for me"

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (alanconway) [reply test/e2e/logforwarding/fluent/fluent_test.go:65]: This line is starting a goroutine to create a recever using t.Client.
go func() {...}() is how you start a goroutine, so 
g.Go(func() {...}) starts a gouroutine managed by the group g. This is syntactically pretty close, and the same pattern as the errgroup package which is a proposed Go extension https://godoc.org/golang.org/x/sync/errgroup.
ExpectOK() is shorthand for Expect(x).To(Succeed()) which *really* annoys me for unnecessry typing in Ginkgo. 
Which part is problematic?

File: test/e2e/logforwarding/fluent/fluent_test.go
Comment (alanconway) [reply test/e2e/logforwarding/fluent/fluent_test.go:75]: The goroutine that starts at line 75 is going to block until it reads some expected log output (or the read times out).
Following this (not shown above) are two goroutines that will block reading on files that shoud *NOT* get any data. I pass this context to those Read() methods. If Read()s succeed that's an error - I wasn't expecting data there.
Once the first groutine succeeds, it cancels the others, otherwise the test will be stuck for some overly-long timeout waiting for data that won't come.
I'm not sure how to express that more clearly - context.Context is the standard way to cancel goroutines if their work is no longer needed. "cancel()" is the name the context library uses.

**Pull request merged**
